Controller/Gateway/MoneroPayment.php

```diff
@@ -330,63 +336,129 @@ public function retriveprice($currency)
-            setcookie('payment_id', $payment_id, time() + 2700);
+        if (!isset($_COOKIE['xmr_subaddress'])) {
+            // Generate subaddress
+            $xmr_subaddress = $this->monero_daemon->create_address(0);
+            setcookie('xmr_subaddress', $xmr_subaddress['address'], time() + 2700);
+        } else {
+            $xmr_subaddress = ['address' => $_COOKIE['xmr_subaddress']];
         }
-        else
-            $payment_id = $_COOKIE['payment_id'];
-        return $payment_id;
+        return $xmr_subaddress['address'];
     }
-    
+
     public function changeto($amount, $currency)
     {
         $rate = $this->retriveprice($currency);
         $price_converted = $amount / $rate;
         $converted_rounded = round($price_converted, 11); //the Monero wallet can't handle decimals smaller than 0.000000000001
-            return $converted_rounded;
+        return $converted_rounded;
+    }
+
+    protected function check_payment_rpc($subaddress)
+    {
+        $txs = array();
+        $address_index = $this->monero_daemon->get_address_index($subaddress);
+        if(isset($address_index['index']['minor'])){
+          $address_index = $address_index['index']['minor'];
+        }
+        else {
+          return $txs;
+        }
+        $payments = $this->monero_daemon->get_transfers(array( 'in' => true, 'pool' => true, 'subaddr_indices' => array($address_index)));
+        if(isset($payments['in'])) {
+          foreach($payments['in'] as $payment) {
+              $txs[] = array(
+                  'amount' => $payment['amount'],
+                  'txid' => $payment['txid'],
+                  'height' => $payment['height']
+              );
+          }
+        }
+        if(isset($payments['pool'])) {
+          foreach($payments['pool'] as $payment) {
+              $txs[] = array(
+                  'amount' => $payment['amount'],
+                  'txid' => $payment['txid'],
+                  'height' => $payment['height']
+              );
+          }
+        }
+        return $txs;
     }
-    
-    public function verify_payment($payment_id, $amount)
+
+    public function verify_payment($payment_id, $amount, $num_confirmations)
     {
-        $message = "We are waiting for your payment to be confirmed";
+        $message = "We are waiting for your payment.";
         $amount_atomic_units = $amount * 1000000000000;
-        $get_payments_method = $this->monero_daemon->get_payments($payment_id);
-        if (isset($get_payments_method["payments"][0]["amount"]))
-        {
-            if ($get_payments_method["payments"][0]["amount"] >= $amount_atomic_units)
-            {
+        $total_received = 0;
+
+        // Fetch transactions for the given payment subaddress
+        $txs = $this->check_payment_rpc($payment_id);
+
+        // If num_confirmations is 0, simply check if payment has been received
+        if ($num_confirmations == 0) {
```

Comment:
Good, I'm thinking about we might want to show something in the settings if num_confirmations == 0. Also, I don't see any check for num_confirmations < 0. Are we sure nobody can actually set that value?

Reply:
added more info and validation validate-zero-or-greater so it looks like: 
![lowerthanzero](https://github.com/user-attachments/assets/30d0b7b1-b950-45fe-9b2e-336b7502dc3f)